cpp/test/sg/fil_test.cu

```diff
@@ -117,6 +124,39 @@ __global__ void nan_kernel(float* data, const bool* mask, int len, float nan)
 
 float sigmoid(float x) { return 1.0f / (1.0f + expf(-x)); }
 
+void hard_clipped_bernoulli(
```

Comment:
Why is this function needed?

Reply:
the default RNG produces values outside the uniform distribution boundaries, leading to 0 probabilities still making categorical features and nodes. This leads to altering the existing tests, where now, we're testing categorical nodes.
Also, this doesn't let us exercise some boundary conditions (e.g. no NANs, or all categories match, or none do).